capnp-rpc-net/vat.ml

```diff
@@ -1,83 +1,75 @@
-open Lwt.Infix
+open Eio.Std
 open Capnp_rpc.Std
 
 module Log = Capnp_rpc.Debug.Log
 
 module ID_map = Auth.Digest.Map
 
-module Make (Network : S.NETWORK) (Underlying : Mirage_flow.S) = struct
+module Make (Network : S.NETWORK) = struct
   module CapTP = CapTP_capnp.Make (Network)
 
   let hash = `SHA256 (* Only support a single hash for now *)
 
-  type connection_attempt = (CapTP.t, Capnp_rpc.Exception.t) result Lwt.t
+  type connection_attempt = (CapTP.t, Capnp_rpc.Exception.t) result Eio.Promise.t
 
   type t = {
+    sw : Eio.Switch.t;
     network : Network.t;
-    switch : Lwt_switch.t option;
     secret_key : Auth.Secret_key.t Lazy.t;
     address : Network.Address.t option;
     restore : Restorer.t;
     tags : Logs.Tag.set;
-    connection_removed : unit Lwt_condition.t;   (* Fires when a connection is removed *)
+    connection_removed : Eio.Condition.t;        (* Fires when a connection is removed *)
     mutable connecting : connection_attempt ID_map.t; (* Out-going connections being attempted. *)
     mutable connections : CapTP.t ID_map.t;      (* Accepted connections *)
     mutable anon_connections : CapTP.t list;     (* Connections not using TLS. *)
   }
 
-  let create ?switch ?(tags=Logs.Tag.empty) ?(restore=Restorer.none) ?address ~secret_key network =
-    let t = {
+  let create ?(tags=Logs.Tag.empty) ?(restore=Restorer.none) ?address ~sw ~secret_key network =
+    Fiber.fork_daemon ~sw Capnp_rpc.Leak_handler.run;
+    {
+      sw;
       network;
-      switch;
       secret_key;
       address;
       restore;
       tags;
-      connection_removed = Lwt_condition.create ();
+      connection_removed = Eio.Condition.create ();
       connecting = ID_map.empty;
       connections = ID_map.empty;
       anon_connections = [];
-    } in
-    Lwt_switch.add_hook switch (fun () ->
-        let ex = Capnp_rpc.Exception.v ~ty:`Disconnected "Vat shut down" in
-        ID_map.bindings t.connections |> Lwt_list.iter_p (fun (_, c) -> CapTP.disconnect c ex) >>= fun () ->
-        t.connections <- ID_map.empty;
-        Lwt_list.iter_p (fun c -> CapTP.disconnect c ex) t.anon_connections >|= fun () ->
-        t.anon_connections <- [];
-        ID_map.iter (fun _ th -> Lwt.cancel th) t.connecting;
-        t.connecting <- ID_map.empty;
-      );
```

Comment:
There's no need to clean up the connections now, because each connection has an owning daemon fiber that will close it when it ends.